provider/spark.go

```diff
@@ -944,40 +972,91 @@ func (s *S3Materialization) NumRows() (int64, error) {
 }
 
 func (s *S3Materialization) IterateSegment(begin, end int64) (FeatureIterator, error) {
-	stream, err := s.store.RowStreamFromSelectQuery(s.Key, fmt.Sprintf("SELECT * FROM S3Object WHERE row_number > %d AND row_number <= %d", begin, end))
+	stream, err := s.store.ResourceStreamConv(s.Key)
 	if err != nil {
 		return nil, err
 	}
+	for i := int64(0); i < begin; i++ {
+		_ = <-stream
+	}
 	return &S3FeatureIterator{stream: stream, maxIdx: (end - begin)}, nil
 }
 
 type S3FeatureIterator struct {
-	stream chan []byte
+	stream chan interface{}
 	cur    ResourceRecord
 	err    error
 	curIdx int64
 	maxIdx int64
 }
 
-// expected format is "<entity(string)>,<value(interface{})>,<timestamp(int64)>"
-func featureCSVToResource(csv string) (ResourceRecord, error) {
-	values := strings.Split(csv, ",")
-	entity := string(values[0])
-	value := reflect.ValueOf(values[1]).Interface()
-	timeStampMilli, err := strconv.Atoi(values[2])
-	if err != nil {
-		return ResourceRecord{}, err
+func convIndirectToVal(value interface{}) interface{} {
+
+	if value == nil {
+		return nil
+	}
+	switch v := value.(type) {
+	case *string:
+		if v == nil {
+			return nil
+		}
+		return *v
+	case *int:
+		return *v
+	case *int32:
+		if v == nil {
+			return nil
+		}
+		return int(*v)
+	case *int64:
+		return *v
+	case *float32:
+		return *v
+	case *float64:
+		return *v
+	case *bool:
+		return *v
+	}
+	return nil
+}
+
+func featureStructToResource(row interface{}) (ResourceRecord, error) {
+	rowVal := reflect.ValueOf(row)
+	if rowVal.NumField() < 3 {
```

Comment:
Is this always 3 or could it be 2?

Reply:
it should always be 3 even if timestamp is not declared in feature def. It will just make a default UTC 0 timestamp